From 5c5a7073aa191f18bee97a9a49fc65242a719d21 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 01 Jul 2021 16:26:33 +0300
Subject: [PATCH] [SW-47213] pci: avoid accessing hw when it's not ready

In some hosts the PRSTn signal might be asserted for a long time.
While it's asserted some hw block are held in reset and must not
be accessed.
after deassertion of the PRSTn signal the HW is performing
initialization sequence during which it's not advisable to access
it as well. (quoting Haim Revivo here)

Hence the sequense of the pci monitoring by preboot should be:
while PRSTn is low, there is no point reading any status regs. in fact it's
dangerous and might cause cpu exception.
after deassertion of the PRSTn we wait for the first link establishment,
that's our indication all the clocks are stable and HW FSMs operational.
and then we start monitoring the rest of the events (flr, hot reset,
link state, etc..)

In order to avoid duplication and race conditions, CPU1 is responsible
for the first stage of PRSTn and link monitoring. once everything is
done a flag is passed to CPU0 via a SP register.

Change-Id: I4e7fe397c19c3061d1e11b875173dca10bbd49ef
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/gaudi/common.c b/gaudi/common.c
index 068a94e..06001f7 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -246,6 +246,8 @@
 }
 static void preboot_pcie_monitoring(void)
 {
+	regmap_writel(PCIE_NOT_INITIALIZED, mmPCIE_STATE);
+
 	/* Wait for PRSTn */
 	while (!regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN))
 		;
@@ -283,6 +285,7 @@
 
 	/* W/A H/W bug that may lead to have a wrong mask (0x1) in LTSSM */
 	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
+	regmap_writel(PCIE_INITIALIZED, mmPCIE_STATE);
 }
 
 void preboot_secondary_cpu_stop(void)
@@ -572,3 +575,8 @@
 
 	return 0;
 }
+
+bool is_pcie_initialized(void)
+{
+	return !!regmap_readl(mmPCIE_STATE);
+}
diff --git a/gaudi/common.h b/gaudi/common.h
index 4ec9ec4..48f59f4 100644
--- a/gaudi/common.h
+++ b/gaudi/common.h
@@ -67,6 +67,7 @@
 int gaudi_hbm_init(void);
 u32 preboot_get_vendor_id_device_id(void);
 void preboot_set_hbm_init(bool is_init);
+bool is_pcie_initialized(void);
 
 #endif /* __ASSEMBLY__ */
 #endif /* COMMON_GAUDI_H_ */
diff --git a/include/preboot.h b/include/preboot.h
index cf1bde4..73751b7 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -80,6 +80,11 @@
 	FIT_SEC_SIGN_FAILED,
 };
 
+enum pcie_init_status {
+	PCIE_NOT_INITIALIZED = 0,
+	PCIE_INITIALIZED
+};
+
 struct hl_nic_en {
 	bool is_valid;
 	uint32_t bitmask;
@@ -202,6 +207,7 @@
 
 void preboot_device_id_init(void);
 u32 preboot_device_id_get(void);
+bool is_pcie_initialized(void);
 
 int preboot_set_sram_sf_access(bool is_enable);
 
diff --git a/preboot/module_pci.c b/preboot/module_pci.c
index 0b3c0ad..7dbd873 100644
--- a/preboot/module_pci.c
+++ b/preboot/module_pci.c
@@ -31,7 +31,7 @@
 	.smlh_link_reg_val = 1,
 	.rdlh_link_reg_val = 1,
 	.link_is_up = false,
-	.mem_enabled_prev = false,
+	.mem_enabled_prev = false
 };
 
 __weak void pci_dev_specific_setup(void) {}
@@ -46,6 +46,7 @@
 }
 
 __weak void hl_pci_iatu_init(void) {}
+__weak bool is_pcie_initialized(void) {return true; }
 
 void module_pci_setup(void)
 {
@@ -55,10 +56,17 @@
 
 __weak enum cmd_check_res dev_specific_module_pci_check(void)
 {
-	pci_sts.flr_reg_val = regmap_readl(mmPCIE_AUX_FLR_INT);
-	pci_sts.hot_rst_reg_val = regmap_readl(mmPCIE_CORE_HOT_RST) & BIT(0);
+	/*
+	 * don't read pci status regs until PRSTn is deasserted and pci blocks
+	 * are functional
+	 */
+	if (!is_pcie_initialized())
+		return NOT_FOUND;
+
 	pci_sts.rdlh_link_reg_val = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
 	pci_sts.smlh_link_reg_val = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
+	pci_sts.flr_reg_val = regmap_readl(mmPCIE_AUX_FLR_INT);
+	pci_sts.hot_rst_reg_val = regmap_readl(mmPCIE_CORE_HOT_RST) & BIT(0);
 
 	if (pci_sts.flr_reg_val == 1 || pci_sts.hot_rst_reg_val == 1)
 		return FOUND_FAST_RETURN;
